Quarkus Configuration Disambiguation strategy for environment variables fails in 3.17.0+

### Describe the bug

[The Quarkus configuration reference guide](https://quarkus.io/guides/config-reference#environment-variables) describes how to disambiguate configuration properties with dynamic segments. The recommended approach is to check in the expected configuration key with a value left empty to enable it to be configured via an environment variable.

### Expected behavior

The server can start/tests can run without providing an optional entry for a dynamic configuration value.



### Actual behavior

This behavior breaks on an upgrade from 3.16.4 to 3.17.0. I tested 3.18.3 and 3.19.0.CR1 to ensure this hasn't been fixed yet. 


It fails with the following error in a Quarkus component test:

```java
Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00040: The config property com.ryandens.deployments."fake-deployment-name-1" is defined as the empty String ("") which the following Converter considered to be null: io.smallrye.config.Converters$BuiltInConverter
io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00040: The config property com.ryandens.deployments."fake-deployment-name-1" is defined as the empty String ("") which the following Converter considered to be null: io.smallrye.config.Converters$BuiltInConverter
	at io.smallrye.config.SmallRyeConfig.buildMappings(SmallRyeConfig.java:139)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:93)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:773)
	at io.quarkus.test.component.QuarkusComponentTestExtension.startContainer(QuarkusComponentTestExtension.java:454)
	at io.quarkus.test.component.QuarkusComponentTestExtension.beforeEach(QuarkusComponentTestExtension.java:237)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: java.lang.NullPointerException: Cannot invoke "java.util.List.iterator()" because "injectedParams" is null
		at io.quarkus.test.component.QuarkusComponentTestExtension.destroyDependentTestMethodParams(QuarkusComponentTestExtension.java:349)
		at io.quarkus.test.component.QuarkusComponentTestExtension.afterEach(QuarkusComponentTestExtension.java:249)
		... 2 more

```


A similar error occurs with the Quarkus test
```java


java.lang.RuntimeException: Failed to start quarkus
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:611)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:695)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:305)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:224)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:578)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:628)
	... 1 more
Caused by: io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00040: The config property com.ryandens.deployments."fake-deployment-name-1" is defined as the empty String ("") which the following Converter considered to be null: io.smallrye.config.Converters$BuiltInConverter
	at io.smallrye.config.SmallRyeConfig.buildMappings(SmallRyeConfig.java:139)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:93)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:773)
	at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
	at io.quarkus.runtime.generated.Config.createRunTimeConfig(Unknown Source)
	at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
	... 8 more


Failed to start quarkus
java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:305)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:224)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:578)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:628)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.smallrye.config.ConfigValidationException: Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00040: The config property com.ryandens.deployments."fake-deployment-name-1" is defined as the empty String ("") which the following Converter considered to be null: io.smallrye.config.Converters$BuiltInConverter
	at io.smallrye.config.SmallRyeConfig.buildMappings(SmallRyeConfig.java:139)
	at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:93)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:773)
	at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
	at io.quarkus.runtime.generated.Config.createRunTimeConfig(Unknown Source)
	at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
	... 8 more


```



### How to Reproduce?

1. Clone the reproducer: https://github.com/ryandens/quarkus-config-disambiguation-reproducer
2. Run the build, seeing the error in the test report/build scan for the `QuarkusComponentTest`
3. Run the build, seeing the error in the test report/build scan for the `QuarkusTest`
4. Run `./gradlew :lib:quarkusDev`, seeing that the app fails to start up and serve the API.

### Output of `uname -a` or `ver`

Darwin mac

### Output of `java -version`

openjdk version "21.0.6" 2025-01-21 LTS

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.18.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

8.12.1

### Additional information

_No response_

--- a/docs/src/main/asciidoc/config-reference.adoc
+++ b/docs/src/main/asciidoc/config-reference.adoc
@@ -97,31 +97,9 @@ property names together.
 
 [source,properties]
 ----
-# value cannot be left empty
-quarkus.datasource."datasource-name".jdbc.url=value
+# value can be left empty but must be supplied by another source at runtime (or be an optional)
+quarkus.datasource."datasource-name".jdbc.url=
 ----
-
-If the value is not known until runtime, the key name can still be provided using the `io.smallrye.config.WithKeys` annotation.
-
-[source,java]
-----
-@ConfigMapping(prefix = "com.example")
-public interface ExampleConfig {
-
-    @WithKeys(KeySupplier.class)
-    Map<String, String> deployments();
-
-
-    class KeySupplier implements Supplier<Iterable<String>> {
-        @Override
-        public Iterable<String> get() {
-            return List.of("fake-deployment-name-2", "fake-deployment-name-1");
-        }
-    }
-}
-----
-
-However, a values for these keys **must** then be provided at runtime. Configuration key disambiguation is not supported for optional properties.
 
 [source,bash]
 ----

--- a/docs/src/main/asciidoc/config-reference.adoc
+++ b/docs/src/main/asciidoc/config-reference.adoc
@@ -91,15 +91,37 @@ configuration name to its most likely dotted format. This works fine for fixed c
 names that contain dynamic segments. In this case, Quarkus is unable to determine if `DATASOURCE_NAME` should be
 converted to `datasource.name` or `datasource-name` (or any other special character separator).
 
-For this reason, such properties always require their dotted version name in another source (the value can be left
-empty) to disambiguate the Environment Variable name. It will provide additional information to perform a two-way
-conversion and match the property names together.
+For this reason, such properties always require their dotted version name in another source to disambiguate the
+Environment Variable name. It will provide additional information to perform a two-way conversion and match the
+property names together.
 
 [source,properties]
 ----
-# value can be left empty
-quarkus.datasource."datasource-name".jdbc.url=
+# value cannot be left empty
+quarkus.datasource."datasource-name".jdbc.url=value
 ----
+
+If the value is not known until runtime, the key name can still be provided using the `io.smallrye.config.WithKeys` annotation.
+
+[source,java]
+----
+@ConfigMapping(prefix = "com.example")
+public interface ExampleConfig {
+
+    @WithKeys(KeySupplier.class)
+    Map<String, String> deployments();
+
+
+    class KeySupplier implements Supplier<Iterable<String>> {
+        @Override
+        public Iterable<String> get() {
+            return List.of("fake-deployment-name-2", "fake-deployment-name-1");
+        }
+    }
+}
+----
+
+However, a values for these keys **must** then be provided at runtime. Configuration key disambiguation is not supported for optional properties.
 
 [source,bash]
 ----

